Add Basic._exec_constructor_hook

mcpl-sympy opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #18768

#### Brief description of what is fixed or changed
Added `Basic._exec_constructor_hook`, and `dispatcher` to `Add`, `Mul` and `Pow`.

#### Other comments
This hook is applied to `Add`, `Mul`, and `Pow` to remove the hard-coding. It can be applied to other classes.  
For this, `dispatcher` attribute is added to `Add`, `Mul` and `Pow`. This returns the hook function of given argument.  
Currently, different arguments with different hooks are not supported.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- core
    - `singledispatch` is introduced to `Add`, `Mul` and `Pow` for constructor hooks.
    - `_exec_constructor_postprocessors` is deprecated.
<!-- END RELEASE NOTES -->

JSS95 committed:
Add Basic._exec_constructor_preprocessor

JSS95 committed:
Change matrix's postprocessor to preprocessor

JSS95 committed:
Apply preprocessor to Add, Mul and Pow

JSS95 committed:
Fix bug in preprocessor

JSS95 committed:
Fix bug in preprocessor

JSS95 committed:
Fix bug and change test name

JSS95 committed:
Add hack to matexpr's preprocessors

This also makes one XFAILed test to pass.

JSS95 committed:
Add tests

oscarbenjamin commented:
I think there needs to be more discussion before introducing preprocessors. I don't like the post-processors as we shouldn't need this in core. I think that it would be better if this logic was all handled in subclasses without needing any code in the core

mcpl-sympy commented:
@oscarbenjamin 
I strongly suggest that this logic needs to be handled in core.  
There are already many `Add`-like and `Mul`-like objects in sympy. `MatAdd` in matrices module, `TensAdd` in tensor module, `VectorAdd` in vector module... and who knows how many more in the future?  

It would be absurd to take all these classes in mind and do the type checking every time like  

```python
if any(isinstance(i, Vector) for i in args):
    return VectorAdd(*args)
elif any(isinstance(i, TensExpr) for i in args):
    return TensorAdd(*args)
elif
...
```

Current operator overriding (`__add__`, `__radd__`, etc) in sympy checks `_op_priority` of the arguments and calls the operation classes (again, i.e. `Add`, `Mul`, `Pow`, ...) of the higher one.  
However, this is quite restricted because it cannot pass any keyword arguments to the result. One eventually needs to use the operation classes explicitly, and the best way to do it will be to forget about all the subclasses and just use `Add`, `Mul`, and `Pow` as constructor - let sympy choose the right subclass.

mcpl-sympy commented:
Plus, current 'postprocessor' approach is very precarious. Not only it cannot pass the keyword arguments, it forces the arguments to be processed by `Add.flatten` and `Add._from_args` and etc.  

Let's say one - maybe a sympy contributer in the future - defined `NewExpr` class, which results `NewAdd` class by addition. Even if `NewAdd` class has its own `flatten` and `_from_args` method, the developer should make sure that `NewAdd` passes `Add.flatten` and `Add._from_args` without any error in order to use `Add(newexpr1, newexpr2)`. This surely will be a nuisance!

oscarbenjamin commented:
> It would be absurd to take all these classes in mind and do the type checking every time like

I agree but I don't like either the pre- or post-processor approach. I also don't like depending on `op_priority` for anything other than deciding the order of operations in operators.

We need a better system for handling this in general. For that there needs to be some discussion and the discussion should begin by clearly identifying what problem these hooks are trying to solve and considering alternative ways of solving the problems.

asmeurer commented:
Can you give a rundown of how this works and how it is different from what is there now?

JSS95 committed:
Add docstring

mcpl-sympy commented:
@asmeurer

preprocessor
---
When arguments and options are passed to constructor (e.g. running 'Add(x, y)'), they are passed to this method.  Then, it compares ``_op_priority`` attribute of the arguments. If any superclass of argument with highest ``_op_priority`` can be found in ``Basic._constructor_preprocessor_mapping``, its preprocessor is used to return the object.  
If ``preprocess=False`` option is passed, the arguments are not preprocessed.

Example:
`Symbol._op_priority` is `10`, and `MatrixSymbol._op_priority` is `11`. Therefore, with `x = Symbol('x')` and `A = MatrixSymbol('A', 2,2)`, `MatrixSymbol`'s preprocessor is selected when 'Mul(x,A)' is run. Since the preprocessor returns `MatMul` in this case, `(x, A)` is passed to `MatMul` and `MatMul(x,A)` is returned without generating `Mul` object.

postprocessor
---
Unlike preprocessor, constructor object is first generated, then transformed to the subclass by postprocessors. Every arguments have different postprocessors, every postprocessors will be applied.  

Example:  
With `x = Symbol('x')` and `A = MatrixSymbol('A', 2,2)`, running 'Mul(x,A)' actually generates `Mul(x,A)` object. Then, postprocessor of `A` is applied to `Mul(x,A)`, returning `MatMul(x,A)` object.

asmeurer commented:
So, let me see if I have this right. The main differences are

- It works as a preprocessor instead of postprocessor, so that if one exists, then the `flatten` methods will not be called.
- It uses `_op_priority` to determine priority. The current postprocessors just go with the first one found by args order.

Are both cases needed?

Questions

- what if you want to return a Mul, rather than a subclass? In that case, you do want to call flatten. 
- How does it handle the case when there are multiple preprocessors?

asmeurer commented:
This is a use-case to consider for where you want to extend the `flatten` algorithm rather than returning a Mul subclass https://github.com/sympy/sympy/pull/13161

mcpl-sympy commented:
@asmeurer 
1. There are two ways to make 'Mul(A,B)' to return `Mul(A,B)`.
- Define no preprocessor at all. If no preprocessor is defined for any arguments, (case for most core classes), `Mul` is constructed and returned normally.
- Preprocessor can return `Mul` instance, if needed. In this case, 'Mul(A,B, preprocess=False)', should be used inside the preprocessor to avoid infinite recursion.
- If only Mul-wise flattened arguments are needed (not `Mul` object), using `Mul.flatten` method in the preprocessor will be a good idea.

2. If you mean by the case where `A` and `B` have different preprocessor while having same `_op_priority`, this should be avoided. But if needed, it can be handled by adding type checking for `B` in `A`'s preprocessor and `A` for `B`'s.


For the case in #13161: `AccumBounds._op_priority` is 11, which collides with `MatrixSymbol`'s. Which one should be higher, do you suggest? I will add preprocessor for `AccumBounds` and push to this PR.

mcpl-sympy commented:
By the way, I think `_op_priority` of every classes should be documented. I will make a page in wiki.

JSS95 committed:
Merge branch 'master' into preprocessor

JSS95 committed:
Apply preprocessor to add and mul of AccumBounds

JSS95 committed:
Delete unnecessary AccumBounds.__sub__ method

mcpl-sympy commented:
@asmeurer
I introduced preprocessor to AccumBounds. Please check it.

oscarbenjamin commented on sympy/calculus/util.py:
```diff
@@ -1649,3 +1552,129 @@ def union(self, other):
 # setting an alias for AccumulationBounds
 AccumBounds = AccumulationBounds
+
+def get_preprocessor(cls):
```
I don't understand the purpose of this function that encloses the two others that are the actual functions that we want...

mcpl-sympy replied:
Frankly speaking, `AccumBounds` does not need preprocessor since its behavior is already hard-coded in `Add.flatten`. I pushed the last three commits to show that with preprocessor, any class can use `Add` and `Mul` as constructor without modifying`flatten`.  

Again, the reason why I am proposing the preprocessor is to let user define his/her own class and make it compatible with `Add` or `Mul`, without meddling their `__new__`, `flatten` or `_from_args`.

mcpl-sympy replied:
I referred to matrix module's `get_postprocessor` function and used its design. Should I discard this wrapper function? I can easily do it.

asmeurer replied:
I believe the matrices one is written like this because the two processors were so similar that it made sense to combine them into a single function. It looks like that isn't the case here, as the Add and Mul parts don't share any code.

mcpl-sympy replied:
Fixed it.

mcpl-sympy commented:
Frankly speaking, `AccumBounds` does not need preprocessor since its behavior is already hard-coded in `Add.flatten` and `Mul.flatten`. I pushed these last three commit to show that one can use preprocessor to make `AccumBounds` compatible with `Add` and `Mul`, without adding any more lines in their `flatten` methods.

Again, the reason why preprocessor is needed is to let user define new class and make it compatible with `Add` and `Mul`, without modifying their methods anymore.

oscarbenjamin commented:
> Frankly speaking, `AccumBounds` does not need preprocessor since its behavior is already hard-coded in `Add.flatten` and `Mul.flatten`.

Yes but it shouldn't be hard-coded in the core flatten methods so it is good to break it out and use the pre/post-processor approach. I still think a better approach is needed in general but at least we should not have special cases hard-coded into the flatten methods.

JSS95 committed:
Spilt get_preprocessor of AccumBounds

JSS95 committed:
Apply preprocessor to TensExpr

There was a bug that two TensorHeads with different `comm` are
regarded to be equal, thus resulting an error when being cached.
Fixed this by adding `_hashable_content` method to TensorHead.

JSS95 committed:
Add tests and fix bug

JSS95 committed:
Make Pow with TensExpr always evaluated

Improved code quality as well.

mcpl-sympy commented:
@asmeurer @oscarbenjamin 
I added preprocessor to TensExpr as well. This removes hard-coding in `Add.flatten`, and also makes `TensExpr` compatible with `Mul` and `Pow`. Please review it!

mcpl-sympy commented:
I will resolve the test failure..

JSS95 committed:
Restore operator methods of TensorExpr

Since Expr's operator method raises error with MutableMatrix,
these methods are needed for TensorExpr to work with physics module.

mcpl-sympy commented:
@asmeurer @oscarbenjamin 
I added preprocessor to `TensExpr` as well. It removes the hard-coding in `Add.flatten`, and also makes it compatible with `Mul` and `Pow`. Please review it!

asmeurer commented:
Would be good to get @Upabjojr's thoughts on this, assuming he has the time to take a look.

mcpl-sympy commented:
@asmeurer  @oscarbenjamin  @Upabjojr 
Should there be any more suggestions, please let me know.

oscarbenjamin commented:
This looks okay to me

czgdp1807 commented:
@asmeurer @Upabjojr Any thoughts on this PR? Can we merge this?

mcpl-sympy commented:
@czgdp1807 @asmeurer @Upabjojr 
I am thinking of applying this to other classes such as `Order`, removing the hard coding of this in `Add` and `Mul`.
Also, classes other than `Add`, `Mul` and `Pow` - perhaps every `Basic` classes` - may use this hook. Then, we can introduce more specified classes, such as `MatrixSubs` which has `shape` attribute, etc.

JSS95 committed:
Change the notion, from 'preprocessor' to 'hook'

JSS95 committed:
Fix test failure

JSS95 committed:
Fix test failure

JSS95 committed:
Add test of AccumBounds' hooks

JSS95 committed:
Apply hooks to vector module

Added tests

JSS95 committed:
Improve code quality

mcpl-sympy commented:
@asmeurer @Upabjojr 
I applied the hook to vector module. Now, `Add(N.i, N.j)` returns `VectorAdd` instance.

JSS95 committed:
Delete MatExpr hard coding in Add.flatten

JSS95 committed:
Remove MatrixExpr hard coding in Mul

JSS95 committed:
Remove unnecessary code in Add

mcpl-sympy commented:
@asmeurer  @oscarbenjamin 
I tried to separate `Order`-related parts in `Add.flatten` and `Mul.flatten`, but these were related in commutative/noncommutative logic and better be handled in the way they are now.

So far, this PR removes hard-coding of `AccumBounds`, `MatrixExpr` and `TensExpr` from `Add` and `Mul`. Also, it allows `sympy.vector` module's vectors to be added and multiplied by `Add` and `Mul`.

It is ready to be merged and I will be grateful if you review it.

oscarbenjamin commented:
I think that this is an improvement but I'm still not entirely happy with the way all this works. It does not really seem scalable to write hooks that deal with more and more combinations of different types including combinations like Order, MatrixSymbol, AccumBounds etc. There needs to be a better way but I'm just not sure what it is...

mcpl-sympy commented:
@oscarbenjamin 
Since this hook uses `_op_priority` to select which one to use, adding new class won't be too complicated.  
Current hooks I made do look tangled. But it is because that's what they were before I extracted them from core. With this PR and #18873, I am planning to make all these jumbled codes simple and clean.  

Without this PR, introducing new class will make `Add` and `Mul` even more perplexing than they are now. If someone comes up with better idea in the future, I'll be happy to adopt it. But for now, this hooks do relieve the complexity and it is crucial to refine all the evaluation logics as proposed in #18873.

mcpl-sympy commented:
@oscarbenjamin 
I created a discussion [page](https://groups.google.com/forum/#!topic/sympy/XNAgvD02_PA) in google groups. I would like to hear your suggestion on this.

asmeurer commented:
For dealing with combinations of classes maybe it would be better to use multipledispatch to define pairwise combinations, which automatically get combined together. 

Although to be honest, most classes that would want to extend Add or Mul do not interact with each other, only with normal Expr classes. Order, MatrixExpr, Vector, and AccumBounds do not know about each other, and adding or multiplying one by the other is, for the most part, mathematically meaningless. In fact, we might even ignore the op_priority issue for now and make it an error if multiple objects have conflicting processors (subclasses that override processors should take priority). Would that cause issues with what is currently here? 

This could change though if we want to take this a step further for Expr classes that are commonly used but have custom add/mul combination logic, such as infinity (but that has other issues such as performance considerations).

oscarbenjamin commented:
If we don't need to think about the interactions then maybe there is a way to approach this using single dispatch rather than a Basic method.

mcpl-sympy commented:
@asmeurer @oscarbenjamin 
OK, so I've come up with this idea.

First, I still think that `_op_priority` should be used to determine which hook will be selected, because it's literally "priority when operated".

Now here is the difference. My prior work couldn't determine which to use when multiple arguments have same `_op_priority`. Instead, I would like to introduce a method, `_operable_classes`. It works like this:  
Suppose we have class `A` and class `B`, both having `11` as `_op_priority`. `A._operable_classes(Add)` returns [`B`], while the result of `B._operable_classes(Add)` does not contain `A`. Then, `Add` on their instances use `A`'s hook.  
If both have each other as operable classes and can't resolve which one to use, hook of the instance which comes first in argment order (`Add(A, B)` or `Add(B, A)`) will be selected.

This is a blueprint, and I will of course add the logic to handle inheritance (e.g. what if `B` does not 'knows' `A`, but knows `A`'s superclass?). How is it?

JSS95 committed:
Introduce new step to determine hook

Basic._operable_classes and Basic._select_hook are introduced.
They will determine the hook when two or more arguments have same
_op_priority.

JSS95 committed:
Fix bug

JSS95 committed:
Fix the logic a bit

JSS95 committed:
Add tests

JSS95 committed:
Add condition for subclass

JSS95 committed:
Fix the contradictory parts

mcpl-sympy commented:
@asmeurer @oscarbenjamin 

I added some more features. These seem a bit complicated, but works alright.  
Now, the hook is selected from the arguments by this logic, with assuming `Add` is used as constructor:

## 1. Check if the hooks are defined at all.
If no hook is found from the arguments, hook is not applied and `Add.__new__` goes normally.  
If only one hook is found, it is used.
If multiple hooks are found, go to next step.

## 2. Check `_op_priority`  
If multiple arguments have each one's own hook, check their `_op_priority`. Hook of the argument with highest `_op_priority` is used.  
If multiple arguments have same `_op_priority`, go to next step.

## 3. Check `_operable_classes`  
In this stage, we pick the argument which 'knows' other arguments the most.  
If `A._operable_classes(Add)` returns `{B}`, it means that `Add(A [, B or B's subclasses])` should use `A`'s hook. With this, every arguments vie with each other to determine which-knows-which by following steps.

### 3-1. Check if the relation is unidirectional

If `A._operable_classes(Add)` contains `B` but `B._operable_classes(Add)` does not contain `A`, it is considered that `A` 'knows' B.  
If both classes' `_operable_classes` contain each other, go to **3-2**.

### 3-2. Check if one contains the other explicitly

Consider we have class `A`, `B`, and `B`'s subclasses `B1`. `A._operable_classes(Add)` contains `B`, and `B1._operable_classes(Add)` contains `A`. In this case, it is considered `B1` knows `A` and not the other way around. It's because `B1._operable_classes` explicitly returns `A` while `A`'s returns `B`'s superclass.  
If `A` and `B` contains each other both explicitly or both implicitly, go to **3-3**.

### 3-3. Check if one is the other's subclass

Consider `B` and its subclass `B2`. Both classes' `_operable_classes` has nothing to do each other. In this case, subclass is preferred so `B2` is selected.

## 4. Refer to the order of arguments

In step **3**, we check the number of known arguments for each arguments. The most-knowing argument's hook is selected. For the case where multiple arguments know exactly same number of other arguments, this step is the final fallback.  
Here, the order of arguments is used to determine which hook to use. `Add(A, B)` uses `A`'s hook, and `Add(B, A)` uses `B`'s hook.  
This step is designed to honor the Python's way where `a+b` returns `a.__add__(b)`, not `b.__radd__(a)`.

asmeurer commented:
I wonder if we can just use multipledispatch directly to implement a lot of this logic.

oscarbenjamin commented:
Perhaps even single dispatch can be used:
https://docs.python.org/3/library/functools.html#functools.singledispatch

mcpl-sympy commented:
@asmeurer  @oscarbenjamin  
I tried with that, but then I was afraid that recursively calling pairwise hooks on every arguments may slow down things. Do we really need to call the hook n times when we add n+1 matrices? That's why I ditched that trial and used this hook-selecting logic.

Well, but I just want to resolve this issue soon. If that's what it takes, I'll try with that again and ping you again.

mcpl-sympy commented:
@asmeurer @oscarbenjamin  
I tried with that, but then I was afraid that recursively calling pairwise hooks on every arguments may slow down things. Do we really need to call the hook n times when we add n+1 matrices? That's why I ditched that trial and used this hook-selecting logic.

Anyway, I'll give it a try and ping you again.

mcpl-sympy commented:
@asmeurer @oscarbenjamin  
I tried with that, but it turned out to be a very bad idea.

The problem is, multiple (or single) dispatch does not support arbitrary-typed, arbitrary numbered arguments. The only way will be, as @asmeurer suggested, implementing pairwise combination.

However, it has two obstacles.

The first obstacle is when we want `evaluate=True`. This makes it unefficient.  
Calling pairwise hook on every arguments may slow down things. When we add n+1 arguments, pairwise hook will be run n times, each cycle including flattening and constructing the object, only to break it apart again in next cycle.

The second obstacle is when we want `evaluate=False`. This makes it ***impossible***.  
`Add` and `Mul` aren't designed for pairwise combination. See this:
```python
>>> from sympy import Add
>>> from sympy.abc import x,y
>>> Add(x,y,x, evaluate=False).args
(x, y, x)
>>> sub_add = Add(x,y, evaluate=False)
>>> Add(sub_add, x, evaluate=False).args
(x + y, x)
```
The second `Add` is pairwisely combined, while the first one is not. You see, discriminating nested-unevaluated-Add and flattened-unevaluated-Add is impossible unless we introduce other keyword arguments and evaluation logics, which will be another monsterous task of job that will take like 20 years and literally deprecate everything.

mcpl-sympy commented:
@asmeurer @oscarbenjamin

Pairwise multiple dispatch is unavailable here. Not only it will make things slow, `Add` and `Mul` are not designed for pairwise combination, especially when it comes to `evaluate=False`.

See this example:
```python
>>> from sympy import Add
>>> from sympy.abc import x,y

>>> Add(x,y,x, evaluate=False).args
(x, y, x)

>>> sub_add = Add(x, y)
>>> Add(sub_add, x, evaluate=False).args
(x + y, x)
```

As you see, it is hard for pairwise combination to discriminate nested-unevaluated-Add and flattened-unevaluated-Add.

mcpl-sympy commented:
If we need multiple dispatch method, perhaps we can use it to 'select the hook', not for the hook itself.  

Ping @asmeurer and @oscarbenjamin. Please let me know your thoughts.

oscarbenjamin commented:
There might be other ways to process the hook. For example it could be a hook using something like single dispatch that is called and somehow indicates which items have or have not been processed so that the hook for matrix expressions only needs to be called once.

I haven't had time to think about this in detail yet. I can think of lots of possible ways but this goes right to the core of sympy so it's something that we need to get right.

mcpl-sympy commented:
@oscarbenjamin  
I decided to use single dispatch method as you suggested, which will make the design far more simple. The drawback for this is that we might not be able to combine different-class arguments, but that's not necessary right now indeed.

JSS95 committed:
Use multipledispatch for hooks

JSS95 committed:
Add See Also section in docstring

JSS95 committed:
Delete redundant code

JSS95 committed:
Change method name

JSS95 committed:
Improve code quality

mcpl-sympy commented:
@asmeurer @oscarbenjamin 

It's done. I introduced single dispatch to register the hook by argument type. Its core feature is in `core.basic`, and you can find out how it's applied in `calculus/util`, `matrices.expressions.matexpr`, `tensor.tensor` and `vector.basisdependent`.  
Please review this.

oscarbenjamin commented on sympy/calculus/tests/test_util.py:
```diff
@@ -545,3 +545,10 @@ def test_issue_16469():
+
+def test_AccumBounds_hooks():
+    a = AccumBounds(1,2)
+    assert Mul(2,a, evaluate=True) == 2*a
```
It would be better if this test asserted what these are equal to like:
```
    assert Mul(2,a, evaluate=True) == 2*a == AccumBounds(2, 4)
```

mcpl-sympy replied:
Fixed it.

oscarbenjamin commented on sympy/calculus/util.py:
```diff
@@ -1649,3 +1552,133 @@ def union(self, other):
+    return Add(result, extra, ignore_hook=True, evaluate=False)
+
+def Mul_hook(*args, **options):
+    evaluate = options.get('evalaute', global_parameters.evaluate)
```
Misspelled evaluate.

A test for this case should be added if there isn't already one.

mcpl-sympy replied:
Got it done. Thanks.

oscarbenjamin commented on sympy/matrices/expressions/tests/test_matexpr.py:
```diff
@@ -587,8 +584,8 @@ def test_MatAdd_postprocessor():
-    # assert a == Add(S.NaN, z) # See the XFAIL above
-    assert isinstance(a, Add)
+    assert a == Add(S.NaN, z)
+    assert a.func == Add
```
We can use `assert unchanged(Add, S.NaN, z)` I think

mcpl-sympy replied:
Fixed it.

oscarbenjamin commented:
It looks like this is using multipledispatch but only in order to implement singledispatch. I think that for single dispatch it is probably better to use the stdlib version:
https://docs.python.org/3/library/functools.html#functools.singledispatch

It seems like `_constructor_hook` is only used in `AssocOp` so I don't see why it needs to be part of Basic. I don't think any of this stuff belongs on Basic.

Is it possible to just have a single dispatch function that is called once-per-arg and normally does nothing but is specialised for the relevant cases like MatrixExpr?

JSS95 committed:
Reflect suggestions

JSS95 committed:
Change to functools.singledispatch

mcpl-sympy commented:
@oscarbenjamin  
I changed to `functools.singledispatch`.

`_constructor_hook` is used `Pow` as well. It needs to be part of `Basic`.

Single dispatch already works once-per-args.

oscarbenjamin commented:
I still don't think any of this belongs in Basic. Actually there are lots of things in Basic that should at least be moved to Expr. For this here I don't see why we can't just add it specifically to Add, Mul, and Pow. Are there any other classes that need to be extended?

mcpl-sympy commented:
@oscarbenjamin

Not any more right now. Then, I'll do as you suggested.

JSS95 committed:
Move _constructor_hook from Basic to AssocOp and Pow

mcpl-sympy commented:
@oscarbenjamin

Got it done. Please review it.

oscarbenjamin commented:
Is the `ignore_hook` parameter needed?

oscarbenjamin commented on sympy/tensor/tensor.py:
```diff
@@ -4199,3 +4201,79 @@ def _expand(expr, **kwargs):
+        raise ValueError('Cannot divide by a tensor')
+
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", category=SymPyDeprecationWarning)
```
Use the `ignore_warnings` or `warns_deprecated_sympy` context managers for this.

mcpl-sympy replied:
Got it done. Thanks!

oscarbenjamin commented on sympy/tensor/tests/test_tensor.py:
```diff
@@ -1968,3 +1968,14 @@ def test_TensorType():
         assert isinstance(S2, TensorType)
+
+@filter_warnings_decorator
+def test_hooks():
```
Why are there warnings coming from the new test?

mcpl-sympy replied:
It's because `_get_valued_base_test_variables` raises deprecation warning. Every other tests which involves tensor's operation use that function, and they have this decorator. I don't know tensor module well so I just followed other tests blindly. Any suggestions?

oscarbenjamin replied:
Perhaps out of scope for this PR but the proper solution is that `_get_valued_base_test_variables` which is part of the testing code here should not be using deprecated code.

mcpl-sympy replied:
Please refer to #15276. Maybe we will need new `TensPow` class, but that needs to be done in another PR so I'll mark this resolved.

JSS95 committed:
Use context manager ignore_warnings in tensor.py

JSS95 committed:
Remove duplicate code and modify test

mcpl-sympy commented:
@oscarbenjamin 

`ignore_hook` is needed to avoid infinite loop. Please see `calculus.util` of this commit for example. This is the case where the hook flattens the `args` and return original class with flattened arguments.

mcpl-sympy commented:
@oscarbenjamin 
Any other comment please? If there aren't, I'd say this is ready to be merged.

oscarbenjamin commented:
Is it possible to use `_from_args` rather than `ignore_hook`?

oscarbenjamin commented:
Would this mechanism also be able to handle `Order` e.g. `Add(x, O(x))`?

oscarbenjamin commented:
Would the `_exec_constructor_postprocessor` method on `Basic` still be needed after the changes here?

mcpl-sympy commented:
@oscarbenjamin 
I planned to use `_from_args` at the first time. But in #5040 I saw a discussion that `_from_args` should be removed from places outside of core, so I eschewed it intentionally.

Also, I tried to apply this to `Order`, but it failed. Please refer to my previous comment:

> I tried to separate Order-related parts in Add.flatten and Mul.flatten, but these were related in commutative/noncommutative logic and better be handled in the way they are now.

Perhaps we can iron out `Add` and `Mul` after this PR, then separate `Order`-handling part using the hook - but that's out of current PR's scope, I reckon.

mcpl-sympy commented:
@oscarbenjamin 
Shall I make `postprocessor` deprecated?

oscarbenjamin commented:
What happens in this example?
```julia
In [22]: X = MatrixSymbol('X', 2, 2)                                                                                              

In [23]: Mul(O(x), y, y**-1, x)                                                                                                   
Out[23]: 
 ⎛ 2⎞
O⎝x ⎠

In [24]: Mul(O(x), X, X**-1, x)                                                                                                   
Out[24]: O(x)⋅x⋅𝕀

In [25]: Mul(x, O(x)) * Mul(X, X**-1)                                                                                             
Out[25]: 
 ⎛ 2⎞  
O⎝x ⎠⋅𝕀
```
On master we get:
```julia
In [1]: X = MatrixSymbol('X', 2, 2)                                                                                               

In [2]: Mul(O(x), y, y**-1, x)                                                                                                    
Out[2]: 
 ⎛ 2⎞
O⎝x ⎠

In [3]: Mul(O(x), X, X**-1, x)                                                                                                    
Out[3]: 
 ⎛ 2  ⎞
O⎝x ⋅𝕀⎠

In [4]: Mul(x, O(x)) * Mul(X, X**-1)                                                                                              
Out[4]: 
 ⎛ 2⎞  
O⎝x ⎠⋅𝕀
```

mcpl-sympy commented:
@oscarbenjamin 
Uh-oh. I'll see to it.

JSS95 committed:
Fix the error in MatMul's hook

mcpl-sympy commented:
@oscarbenjamin 
Current behavior which returns `O(x)⋅x⋅𝕀` is clearly an error. But is master's result right? I think it should return `O(x**2)⋅𝕀`, not `O(x**2⋅𝕀)`.  
Fixed it to return `O(x**2)⋅𝕀` and added a test. If `O(x**2⋅𝕀)` is right result, please tell me and I'll fix again.

JSS95 committed:
Add deprecation warning to postprocessor

JSS95 committed:
Fix circular import

JSS95 committed:
Delete the usage of postprocess from AssocOp and Pow

JSS95 committed:
Delete constructor_postprocessor test

mcpl-sympy commented:
@oscarbenjamin 
I deprecated the postprocessor and deleted its test.

oscarbenjamin commented:
Looks like there's a merge conflict.

I'm still not convinced this is the right solution for extensibility of the core. The fact that it doesn't work for Order makes me think that something else is needed.

mcpl-sympy commented:
@oscarbenjamin 
Conflict's due to deleted file. Maybe I'll leave that file and add warning decorator to the tests.

Extracting `Order` is possible. It's just that `Order` is so entrenched in `flatten` so that it will need some adjustments which can be outside of the scope of PR. Any other attmpt to remove it from `flatten` will face the same issue.

I can apply this to `Order` if you think it's needed in this PR.

oscarbenjamin commented:
It's not about what is needed in this PR. I'm reluctant to make any change unless it looks like it can solve the broader problems in general. Whether this approach does or does not work for Order affects that.

mcpl-sympy commented:
@oscarbenjamin 
I understand. Then, I'll prove that this change can be generally applied by solving `Order` issue. Let me ping you when its' done.

JSS95 committed:
Revert "Delete constructor_postprocessor test"

This reverts commit 78dd9cd78a6b3954f5578dcb1aab8ab0c971955c.

JSS95 committed:
Revert "Delete the usage of postprocess from AssocOp and Pow"

This reverts commit 7f1cdb6c2bd2c6816a74f4c8890f46b6d15c9f40.

JSS95 committed:
Add warn_deprecated_sympy to postprocessor test

JSS95 committed:
Migrating Order out of Add.flatten

JSS95 committed:
Remove Order from Add.flatten

JSS95 committed:
Move Order out of Mul.flatten

oscarbenjamin commented:
You don't need to implement everything in one PR. It just needs to be clear that we have a workable plan. I don't like the current post-processors but unless we are replacing them with something that can work in general it is better to live with them for now.

JSS95 committed:
Remove order-related parts from AssocOp

mcpl-sympy committed:
Merge branch 'master' into preprocessor

JSS95 committed:
Improve code quality

JSS95 committed:
Merge branch 'preprocessor' of github.com:mcpl-sympy/sympy into preprocessor

oscarbenjamin commented:
This SO question seems to be related:
https://stackoverflow.com/questions/62043082/sympy-cannot-evaluate-dot-product-of-metamorphosed-vectors

mcpl-sympy commented:
@oscarbenjamin 
Voila! It's done. `Order` is now extracted from `Add` and `Mul`. Went easier than expected. Don't know why I failed before. I hope this be auspicious in getting this PR accepted. Please review this.

I will check if I can solve the issue that you linked as well.

mcpl-sympy commented:
@oscarbenjamin

That SO question is interesting. It seems that it can be solved by supporting non-evaluated `VectorMul` instance, which makes the issue related to #18873 as well.

oscarbenjamin commented:
I think part of the problem is that factor turns everything into Add/Mul rather than VectorAdd/VectorMul:
```julia
In [33]: from sympy import * 
    ...: from sympy.vector import * 
    ...: N = CoordSys3D('N') 
    ...: x = symbols('x') 
    ...: v = x * N.i + x**2 * N.j                                                                                                 

In [34]: v                                                                                                                        
Out[34]: x*N.i + x**2*N.j

In [35]: type(v)                                                                                                                  
Out[35]: sympy.vector.vector.VectorAdd

In [36]: vf = factor(v)                                                                                                           

In [37]: vf                                                                                                                       
Out[37]: x⋅(i_N + j_N⋅x)

In [38]: type(vf)                                                                                                                 
Out[38]: sympy.core.mul.Mul

In [39]: dot(vf, vf)                                                                                                              
Out[39]: (x⋅(i_N + j_N⋅x))⋅(x⋅(i_N + j_N⋅x))
```
Once that happens there's no way to get back to VectorAdd/VectorMul apart from deconstructing again:
```julia
In [40]: vf2 = vf.args[0] * (vf.args[1].args[0].args[0] * vf.args[1].args[0].args[1] + vf.args[1].args[1])                        

In [41]: vf2                                                                                                                      
Out[41]: x*N.i + x**2*N.j

In [43]: dot(vf2, vf2)                                                                                                            
Out[43]: 
 4    2
x  + x
```
Does this mean that VectorAdd/Mul need to have hooks?

(I'm not suggesting adding those here. Just thought I'd ask about this since it seems related)

JSS95 committed:
Allow power with exponent=1 to vector module

mcpl-sympy commented:
@oscarbenjamin 
Yes, I believe it can be solved by introducing hooks to `VectorAdd` and `VectorMul`.  
These classes already have hook in this PR. I tested this on this branch, and here is a result:

```python
>>> from sympy import *
>>> from symy.vector import *
>>> N = CoordSys3D('N')
>>> x = symbols('x')
>>> v = x * N.i + x**2*N.j
>>> vf = factor(v)
>>> vf
x * N.i + x**2*N.j
>>> type(vf)
sympy.vector.vector.VectorAdd
```

Here, `factor(v)` returns `x * N.i + x**2*N.j`, not `x*(N.i + x*N.j)` because `VectorAdd` and `VectorMul` are always evaulated. If we allow `evaluate=False` to thses classes, perhaps this issue can be solved.

oscarbenjamin commented:
The mechanism here does not seem to guarantee commutativity when mixing terms that have different hooks:
```julia
In [1]: o1 = AccumBounds(1, 2)                                                                                                    

In [2]: o2 = O(1)                                                                                                                 

In [3]: Add(o1, o2)                                                                                                               
Out[3]: O(1)

In [4]: Add(o2, o1)                                                                                                               
Out[4]: <1, 2> + O(1)
```
The order in which hooks are called needs to be well-defined somehow.

mcpl-sympy commented:
@oscarbenjamin 
Well yeah... but I thought you and asmeurer dissuaded me from allowing these combinations.

asmeurer mentioned:
> ... Although to be honest, most classes that would want to extend Add or Mul do not interact with each other, only with normal Expr classes. Order, MatrixExpr, Vector, and AccumBounds do not know about each other, and adding or multiplying one by the other is, for the most part, mathematically meaningless. ...

You mentioned:
> If we don't need to think about the interactions then maybe there is a way to approach this using single dispatch rather than a Basic method.

oscarbenjamin commented:
It seems there is some interaction though. This is the behaviour on master:
```julia
In [1]: o1 = AccumBounds(1, 2)                                                                                                    

In [2]: o2 = O(1)                                                                                                                 

In [3]: o1 + o2                                                                                                                   
Out[3]: <1, 2> + O(1)

In [4]: o2 + o1                                                                                                                   
Out[4]: <1, 2> + O(1)

In [5]: Add(o1, o2)                                                                                                               
Out[5]: O(1)

In [6]: Add(o2, o1)                                                                                                               
Out[6]: O(1)
```
That's inconsistent in a different way. We do need to ensure that addition is always commutative and consistent between `+` and `Add`.

mcpl-sympy commented:
@oscarbenjamin 
My previous attempt to solve this was to use `_op_priority`, and I still think it's a good way. By introducing this to `Add` and `Mul`'s hooks, I can solve this maintaining singledispatch-oriented design. Shall I proceed?

oscarbenjamin commented:
Yes, I think so. That's most likely to be backward compatible since that's how this is currently handled for `__add__` etc.

Does the `_op_priority` differ for all of the classes that define hooks?

oscarbenjamin commented:
Btw tests should be added each time something is fixed (e.g. the last commit for `Vector`).

mcpl-sympy commented:
According to [this](https://github.com/sympy/sympy/wiki/op_priority) wiki page, `AccumBound._op_priority` and `MatrixExpr._op_priority` collides. Also, `TensExpr` and `Vector` collides as well.

This collision itself is an issue, and I think it will be good to solve it here.

Detail of my plan is:

1. When `A.op_priority` > `B._op_priority`, `Add(B,A)` or `Mul(B,A)` returns `A`'s hook. This is congruent to how `B+A` returns `A.__add__(B)`.

2. When two classes collide, check if one is subclass of another. Subclass will have higher priority.

3. If hook still cannot be determined, then the first-coming one is selected.

mcpl-sympy commented:
> Btw tests should be added each time something is fixed (e.g. the last commit for Vector).

Sorry. I will add it right away.

JSS95 committed:
Add vector's pow test

JSS95 committed:
Enhance AssocOp's hook

JSS95 committed:
Fix bug and add test

mcpl-sympy commented:
@oscarbenjamin 
Fixed it. Now, it behaves like this:

```python
In [1]: o1 = AccumBounds(1, 2)                                                                                                    

In [2]: o2 = O(1)                                                                                                                 

In [3]: o1 + o2                                                                                                                   
Out[3]: <1, 2> + O(1)

In [4]: o2 + o1                                                                                                                   
Out[4]: <1, 2> + O(1)

In [5]: Add(o1, o2)                                                                                                               
Out[5]: <1, 2> + O(1)

In [6]: Add(o2, o1)                                                                                                               
Out[6]: <1, 2> + O(1)
```

This is because `AccumBounds._op_priority` is 11 where `Order._op_priority` is 10. Added test and it goes well.

mcpl-sympy commented:
@oscarbenjamin

How do you think of this PR now? If you have any more suggestion please let me know.

eric-wieser commented:
I haven't followed the full discussion here, but this would be of great value to pygae/galgebra, which currently doesn't really integrate with sympy at all due to difficultly. Internally, it has three types of expression that would benefit from this treatment:

* `MvExpr`, which overloads the `^`, `|`, `<<`, and `>>` operators. `MvAdd` obviously wants these overloads too.
* `DiffOpExpr`, with semantics roughly `PartialDiffOp(x) * f(x) == Derivative(f(x), x)`, where we want to overload the left multiplication operator
* `MvDiffOpExpr`, which combines both semantics.

This at first glance looks like a case where the multiple dispatch discussed above is necessary, as `MvExpr * DiffOpExpr` should give `MvDiffOpExpr`. However, this is a case where the giving `MvExpr` higher precedence and letting it know about `MvDiffOpExpr` internally is probably fine.

I'll try to review this PR and discussion properly at some point in future.

oscarbenjamin commented:
I still think a different approach is needed somehow. I'll spell out an idea soon.

Note that the main issue here is not operator overloads which should work okay with sympy (after various fixes to the special methods on Expr in 1.6) provided you have the op_priority set.

The issue is with putting things into Add, Mul, Pow e.g.:
```julia
In [1]: M = MatrixSymbol('M', 2, 2)                                                                                               

In [2]: N = MatrixSymbol('N', 2, 2)                                                                                               

In [3]: Add(M, N)                                                                                                                 
Out[3]: M + N

In [4]: type(_)                                                                                                                   
Out[4]: sympy.matrices.expressions.matadd.MatAdd
```
The reason you would (potentially) want to be able to put things into an Add and then either alter the behaviour of Add or return a different type such as MatAdd is because then it would hopefully work with core routines like expand, collect etc that manipulate Adds and Muls.

I'm wondering though if a better fix is to have functions `sum` and `prod` that do this dispatching so we never have to put objects like `MatrixSymbol` actually into the args of a normal `Add`.

The question is do you need more than binary operators and if so is there ever any need to combine these things in a single Add/Mul. For example do we want to be able to have a vector and a matrix in the same Mul or should that just raise an error?
```julia
In [18]: M = MatrixSymbol('M', 2, 2)                                                                                              

In [19]: i = CoordSys3D('C').i                                                                                                    

In [20]: M*i                                                                                                                      
Out[20]: M*C.i

In [21]: i*M                                                                                                                      
Out[21]: M*C.i

In [22]: type(_)                                                                                                                  
Out[22]: sympy.vector.vector.VectorMul

In [23]: Mul(i, M)                                                                                                                
Out[23]: i_C⋅M

In [24]: type(_)                                                                                                                  
Out[24]: sympy.matrices.expressions.matmul.MatMul
```
Note that multiple dispatch works fine if we have a need for binary dispatching but will also fail if we need n-ary dispatching. Hopefully we can get away with not even needing binary dispatching.

eric-wieser commented:
> I'm wondering though if a better fix is to have functions `sum` and `prod` that do this dispatching so we never have to put objects like `MatrixSymbol` actually into the args of a normal `Add`.

The problem with this approach is that ever occurence of `Add(x, y)` needs to be replaced with `sum(x, y)`, unless you're super sure that `x` and `y` can never be matrices. Pragmatically, it's simpler to just change `Add` to have the proposed semantics, although I suppose it imposes a performance cost sympy-wide

eric-wieser commented:
> The question is do you need more than binary operators and if so is there ever any need to combine these things in a single Add/Mul.

As a contrived example, in a hypothetical sympy-ified galgebra, I want to be able to write
`Add(S.One, MvExpr(...), DiffOpExpr(...))`, and have it give me any one of:
* `MvDiffOpAdd(S.One, MvExpr, DiffOpExpr)
* `MvDiffOpAdd(MvAdd(S.One, MvExpr), DiffOpExpr)
* `MvDiffOpAdd(S.One, MvDiffOpAdd(MvExpr, DiffOpExpr))

mcpl-sympy commented:
@oscarbenjamin 
Why restrict to binary operation while `Add` and `Mul` supports n-ary operation?

mcpl-sympy commented:
Seriously, all the problems that `postprocessor` had can be resolved by this implementation. It passes the keyword arguments to hook, reduces unnecessary steps, deal with multiple type combination, reduces the complexity in `Add` and `Mul` by great degree, and proved to be generally applicable by separating `Order` related logics safely and efficiently. 

Also, this hook can be applied to not only `Add`, `Mul`, and `Pow` but to other classes. as I mentioned in [previous comment](https://github.com/sympy/sympy/pull/18769#issuecomment-631207811), `Derivative` or `Subs` to matrix can return something like `MatrixDerivative` or `MatrixSubs` which is `MatrixExpr` and contains shape information. Its expandability is limitless.

eric-wieser commented on sympy/calculus/util.py:
```diff
@@ -1629,3 +1532,133 @@ def union(self, other):
+
+    extras, accbnds, others = [], [], []
+    for a in args:
+        if isinstance(a, MatrixExpr):
+            extras.append(a)
```
Should this instead be `if not a.is_extended_real`? As written, this looks like exactly the special case you wanted to avoid.

eric-wieser commented on sympy/calculus/util.py:
```diff
@@ -1629,3 +1532,133 @@ def union(self, other):
+            accbnd = a
+        else:
+            accbnd = AccumBounds(
+                Add(accbnd.min, a.min, evaluate=True),
+                Add(accbnd.max, a.max, evaluate=True))
```
Shouldn't this be
```suggestion
                Add(accbnd.min, a.min, evaluate=evaluate),
                Add(accbnd.max, a.max, evaluate=evaluate))
```

eric-wieser commented on sympy/core/operations.py:
```diff
@@ -49,15 +53,74 @@ def __new__(cls, *args, **options):
+
+        hooks = []
+        for a in args:
+            if isinstance(a, cls):  # To find hook in e.g. Add(1, Order(x))
+                func, op_priority, hook = cls._find_hook(*a.args)
```
To be clear, this code path is hit for the inner `Add` in `Add(Add(1, Order(x), 1)`?

oscarbenjamin commented:
> > I'm wondering though if a better fix is to have functions `sum` and `prod` that do this dispatching so we never have to put objects like `MatrixSymbol` actually into the args of a normal `Add`.
> 
> The problem with this approach is that ever occurence of `Add(x, y)` needs to be replaced with `sum(x, y)`, unless you're super sure that `x` and `y` can never be matrices. Pragmatically, it's simpler to just change `Add` to have the proposed semantics, although I suppose it imposes a performance cost sympy-wide

This makes it very hard to write the `Add` class itself. See also https://github.com/sympy/sympy/issues/19429#issuecomment-634612815.

The idea that we can funnel anything we want through Add and money-patch it to produce a different object and then have all the routines like `collect`, `expand` etc just work without modification seems like a pipe dream to me. Work is needed everywhere to be able to support these things. Changing `Add` to `sum` is the easy part - making the code actually work is much harder.

The advantage of having a `sum` function is that it means we can be explicit in the code about what kinds of objects are supported by a given routine and it's clear how to add the support when it is wanted. We would need to go through and decide which functions this actually makes sense for and adapt them and test them to make sure that they work.

Right now what happens in many places in sympy is that you have one set of authors writing code for something like `expand`, `Add` or `Matrix` with one set of expectations and then other sets of authors putting in other random objects that code wasn't designed for. The lack of type-checking means that it sort of works sometimes.

The result is an endless stream of bug reports because things seem to work just enough for someone to try and use them and then discover some or other problem because actually there is no real code implementing what they are trying to do. We fudge the code in one place or another to fix one issue or another but it doesn't change the basic fact that the code is being used in a way that it was never designed for.

This conflict disrupts progress on either side because we end up not being able to improve the core because of all the ways that other code is abusing it and yet things still don't work properly for all the obscure objects that try to alter the behaviour of the core.

The mechanism in this PR is an improvement on the post-processors and cleans up the flatten methods which is good. To me it still feels like a kludge that I would want to eliminate in future though.

mcpl-sympy commented:
@oscarbenjamin 
Okay. Then let's make ourselves clear.  
We all want something that we can funnel down everything. I'm suggesting that we'd use `Add` and `Mul` for that funnel, and your opinion is to use something another (`sum` and `prod`). Did I get it right?

OK, I'll do it then. Instead of `sum` and `prod` whose names are already occupied, what about using the name `add` and `mul`? Will you accept that change if I make them and migrate every features introduced in this PR?

mcpl-sympy commented:
BTW, I don't expect that we'll be able to apply `collect` and `expand` on everything without modifying them. I have never been expecting that. The fact that this PR might help solving #19446 is purely incidental. My purpose has been, and still is, to get rid of `postprocessor` and introduce something better.

oscarbenjamin commented:
> BTW, I don't expect that we'll be able to apply `collect` and `expand` on everything without modifying them. I have never been expecting that.

I don't mean that you in particular expected that. I just presume that this is how "we" ended up down the path of having Add translate to MatAdd etc which is how the hooks got there to begin with.

There's no need to change anything yet until we've discussed and agreed what to do. I'm just giving some of my thoughts but haven't had time to think it through completely. It takes me a long time to review this PR each time you make changes partly because the changes are complex but also because I feel like there is a better way but I'm just not sure.

It might still be a better idea to merge this as it is for now anyway and think of the sum suggestion as a longer term possibility. If you want to take this PR in a very different direction then it's better to open a new PR leaving this one here.

mcpl-sympy commented:
@oscarbenjamin 
I've got time, and I've got zeal. I will make another PR to introduce `sum` and `prod`. It will be better to compare two working branch to see which one is better, rather than doing thought-experiment here. And should you come up with another idea, please let me know and I'll dedicate.

Thank you for reviewing and pondering for this PR so far.

eric-wieser commented:
> If there are two _sum_types T1 and T2 then there is a conflict and it is an error [...] is what I'm not sure about.

For my use case I need more than this - the condition needs to be something like "if there are two sum types, and the types do not explicitly cooperate via [some additional hook]".

oscarbenjamin commented:
> For my use case I need more than this

Can you spell this out in more detail? What are the possibilities that don't work in your use case?

eric-wieser commented:
I'm referring to an above comment, edited with a link to it

oscarbenjamin commented:
I think that if it is possible to write a function that can determine for all of the galgebra objects what the result of an add or multiply should be then there doesn't need any special dispatch in core to make that logic work. It just needs core to call that function that can decide what to do. The reason we need a dispatching mechanism is not so that the mechanism itself would control all of the logic but so that it is possible for different codebases to cooperate while extending the behaviour of the core.

oscarbenjamin commented:
See #14490 for a previous attempt at this.

Upabjojr commented on sympy/core/basic.py:
```diff
@@ -1826,13 +1826,28 @@ def rewrite(self, *args, **hints):
+        After an object (e.g. ``Add(x, y)``) is constructed, this method checks
+        every superclass of its every arguments. If any of the superclasses can be
+        found in ``Basic._constructor_postprocessor_mapping``, its postprocessors are
+        collected. When it is done for every arguments, every postprocessors are
```
Shouldn't _every_ be followed by a singular noun? I would say either _"every postprocessor is ..."_ or _"all postprocessors are ..."_

mcpl-sympy replied:
Oops. My bad...

Upabjojr commented on sympy/calculus/util.py:
```diff
@@ -1629,3 +1532,133 @@ def union(self, other):
+
+    extras, accbnds, others = [], [], []
+    for a in args:
+        if isinstance(a, MatrixExpr):
```
I really think that we should define the behaviours for the association of pairs of types in associative operators. Single dispatch is probably not enough, I think we need double dispatch.

mcpl-sympy replied:
Oops. My bad...

mcpl-sympy replied:
Agreed. I'll try it in #19463.

oscarbenjamin replied:
Multiple dispatch works when you have a fixed number of types you need to dispatch on e.g. in a binary add `a+b` you only ever need to dispatch on `type(a)` and `type(b)`. In a general add 'a+b+c+...' there can be any number. Multiple dispatch doesn't support variadic functions in general.

mcpl-sympy replied:
@oscarbenjamin 
It's true, but I have come to think that with well-designed classes, recursively calling binary dispatcher can solve arbitrary numbered combination.

Previously, in this commit (#18769), this was impossible because `Add` and `Mul` were used as `add` and `mul` constructor, and I had to deal with bunch of back compatibility issues. Instead, please review the recent commit (48b9920) in #19463.

JSS95 committed:
Fix typo
